photosynthesis_metrics/gmsd.py

```diff
@@ -0,0 +1,253 @@
+                 scale_weights: Optional[Union[Tuple[float], List[float]]] = None,
+                 chromatic: bool = False, beta1: float = 0.01, beta2: float = 0.32,
+                 beta3: float = 15.) -> None:
+        super().__init__()
+
+        # Generic loss parameters.
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.data_range = data_range
+        
+        # Values from the paper
+        if scale_weights is None:
+            self.scale_weights = torch.tensor([0.096, 0.596, 0.289, 0.019])
+        else:
+            self.scale_weights = torch.tensor(scale_weights)
+
+        self.chromatic = chromatic
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.beta3 = beta3
+            
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Gradient Magnitude Similarity Deviation (GMSD) as a loss function.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        Returns:
+            Value of GMSD loss to be minimized. 0 <= GMSD loss <= 1.
+        """
+        _validate_input(x=prediction, y=target, scale_weights=None)
+        prediction, target = _adjust_dimensions(x=prediction, y=target)
+
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        if self.data_range == 255:
+            prediction = prediction / 255.
+            target = target / 255.
+            
+        scale_weights = self.scale_weights.to(prediction)
+        ms_gmds = []
+        for scale in range(4):
+            if scale > 0:
+                # Average by 2x2 filter and downsample
+                padding = (prediction.shape[2] % 2, prediction.shape[3] % 2)
+                prediction = F.avg_pool2d(prediction, kernel_size=2, padding=padding)
+                target = F.avg_pool2d(target, kernel_size=2, padding=padding)
+        
+            score = _gmsd(prediction, target, reduction=None)
+            ms_gmds.append(score)
+        
+        # Concat results in different scales and multiply by weight
+        ms_gmds_val = scale_weights.view(1, 4, 1, 1) * (torch.cat(ms_gmds, dim=1) ** 2)
+        # Sum and take sqrt per-image
+        ms_gmds_val = torch.sqrt(torch.sum(ms_gmds_val, dim=-1))
+        
+        score = ms_gmds_val
+        
+        if self.chromatic:
+            assert prediction.size(1) == 3, "Chromatic component can be computed only for RGB images!"
+            
+            # Convert to YIQ color space https://en.wikipedia.org/wiki/YIQ
+            iq_weights = torch.tensor([[0.5959, -0.2746, -0.3213], [0.2115, -0.5227, 0.3112]]).t()
+            prediction_iq = torch.matmul(prediction.permute(0, 2, 3, 1), iq_weights).permute(0, 3, 1, 2)
+            target_iq = torch.matmul(target.permute(0, 2, 3, 1), iq_weights).permute(0, 3, 1, 2)
```

Comment:
The prediction and target are not equal to input ones as avgpooling was performed several times.

Reply:
See MS-GMSD paper, chapter 3.3
> It should be noted that chrominance dissimilarity is only evaluated at Scale M because the spatial resolution to color stimulation is much lower than luminance for vision perception.